### PR TITLE
add deathmatch command for menu; Add low gravity zone

### LINKED ISSUE #686: Deathmatch additions
Some new suggestions to add to DM zones

- [x] A deathmatch with 2 teams instead of FFA
- [x] Regain (some) health upon making a kill.
- [x] Easier way to see list of deathmatches
- [x] Keeping track of kills/death (maybe accuracy/damage)

### LINKED ISSUE #735: /deathmatch
I'm thinking about a new look for /deathmatch instead of showing the text '**_Usage: /deathmatch [zone]_**' to look like /minigames(for e.g.) with much more information about deathmatch zones.

e.g.
**Deathmatch zones*
**_/deathmatch 1  - Running Weapons (RW) - skin hit
/deathmatch 2 - Walking Weapons (WW) - skin hit
/deathmatch 3 - Running Weapons (RW) - lag shot_**

### LINKED ISSUE #745: When leaving deathmatch lagshot is not always removed.
Sometimes when a player uses /deathmatch leave his lag shot setting is not set back to lagcomp.

Might have something to do with lag/desync?

### PR DESCRIPTION
**Pull request by OttoRocket-LVP**
Thank you for making a contribution to [Las Venturas Playground](https://sa-mp.nl/)!

Please take a moment to fill in the questions in this template, to make it easier for our developers
to understand what you're doing.

## What is being changed?
_ - Adding a menu to go to a deathmatch zone.
 - Having a zone with lower gravity (rockets)_

## Why is this being changed?
_To make it easier to select a zone and see what are the settings of specific zones.
It's fun to play with gravity!_

## How are you making the change?
  [X] I have a working check-out of Las Venturas Playground.
  [X] I have included tests for any changed JavaScript code.
  [x] I have tested this change myself on my local server.

### CODE REVIEW

File: javascript/features/death_match/death_match_commands.js [hunk 1]
```diff
@@ -21,9 +23,19 @@ export class DeathMatchCommands {
     }
 
     // The death match command is being used.
-    onDmCommand(player, zone) { 
-        if(zone === null || zone === undefined || typeof zone !== 'number') {
-            player.sendMessage(Message.DEATH_MATCH_INVALID_ZONE, zone);
+    async onDmCommand(player, zone) { 
+        if(isNaN(zone)) {
```
Comment (RussellLVP) [inline javascript/features/death_match/death_match_commands.js]: One of the reasons that I suggested that the verification of input `goToDmZone` does right now should move to `death_match_commands.js`, is because you're now creating a tri-state out of it:

  * `/deathmatch`: displays a dialog
  * `/deathmatch 1`: goes to a zone
  * `/deathmatch 123`: shows an error message

While this is fine on itself, it's inconsistent with all other dialog-based commands that we have, which show the dialog for incorrect input as well.

Code's much easier to reason about when all input validation is done in one place (e.g. the player command), after which the parameters are presumed to be correct unless they're coding errors.

File: javascript/features/death_match/death_match_commands.js [hunk 2]
```diff
@@ -21,9 +23,19 @@ export class DeathMatchCommands {
...
+            const dialog = new Menu('Choose a death match zone.', ['Zone', 'Name', 'Teams', 'Lag shot']);
+            for(const zoneId of this.manager_.validDmZones()) {
+                const location = DeathMatchLocation.getById(zoneId);  
+                dialog.addItem(location.id, location.name, location.hasTeams, location.lagShot,
```
Comment (RussellLVP) [inline javascript/features/death_match/death_match_commands.js]: What does this look like? Do we display "true" and "false" in the menu now?

File: javascript/entities/player.js [hunk 3]
```diff
@@ -60,6 +60,8 @@ export class Player extends Supplementable {
     // Default value for not being in a tam = 255 
     // https://wiki.sa-mp.com/wroot/index.php?title=SetPlayerTeam
     static kNoTeam = 255;
+    // Default gravity value
```
Comment (RussellLVP) [inline javascript/entities/player.js:63]: nit: blank line above here.

File: javascript/features/death_match/death_match_manager.js [hunk 4]
```diff
@@ -102,11 +94,28 @@ export class DeathMatchManger {
...
+        const location = DeathMatchLocation.getById(zone);
+
+        this.playersInDeathMatch_.delete(player);
+        this.playerTeam_.delete(player);
+        
+        player.activity = Player.PLAYER_ACTIVITY_NONE;
+        player.team = Player.kNoTeam;
+        if(!isNaN(location.gravity)) 
+            player.gravity = Player.kDefaultGravity;
+
```
Comment (RussellLVP) [inline javascript/features/death_match/death_match_manager.js:104]: Given #745, should lag shot be in here too?

File: javascript/features/death_match/death_match_manager.test.js [hunk 5]
```diff
@@ -319,4 +310,26 @@ describe('DeathMatchManager', (it, beforeEach) => {
 
         assert.equal(server.textDrawManager.getForPlayer(gunther, textDraw), null);
     });
+
```
Comment (RussellLVP) [inline javascript/features/death_match/death_match_manager.test.js:290]: Why don't we have to test the dialog?

**Review by RussellLVP — CHANGES_REQUESTED** [review]

File: javascript/features/death_match/death_match_commands.js (hunk 2 above)
Comment (OttoRocket-LVP) [reply javascript/features/death_match/death_match_commands.js]: Yes, I assumed this is understandable by the user?
I thught about adding 1/0 but figured that wouldn't imrpove that much.

Maybe a Yes/No would be better if thought not to be understandable by the user?

File: javascript/features/death_match/death_match_manager.test.js (hunk 5 above)
Comment (OttoRocket-LVP) [reply javascript/features/death_match/death_match_manager.test.js:290]: This is part of the deathmatch_command. Not the manager.
There's a test clicking on the frist item going to the first zone.

File: javascript/features/death_match/death_match_commands.js (hunk 1 above)
Comment (OttoRocket-LVP) [reply javascript/features/death_match/death_match_commands.js]: Agreed, makes things easier inded.
Also agreed on validating user input at the command.

File: javascript/features/death_match/death_match_commands.js (hunk 2 above)
Comment (OttoRocket-LVP) [reply javascript/features/death_match/death_match_commands.js]: Discussed with player: "Yes/No" is preferred

File: javascript/features/death_match/death_match_manager.js (hunk 4 above)
Comment (OttoRocket-LVP) [reply javascript/features/death_match/death_match_manager.js:104]: I don't think we want to do it here as this is also passed by when switching from DM zone.

Resulting in it switching to 2 and 0 in the same frame. (What could potentially cause the wrong package to arrive later due to desync I assume?)

I added it to the leave with abuse though. That's where we're missing it.

File: javascript/features/death_match/death_match_commands.js [hunk 6]
```diff
@@ -3,27 +3,55 @@
...
-            player.sendMessage(Message.DEATH_MATCH_INVALID_ZONE, zone);
+    async onDmCommand(player, zone) { 
+        if(isNaN(zone) || !DeathMatchLocation.hasLocation(zone)) {
+            const dialog = new Menu('Choose a death match zone.', ['Zone', 'Name', 'Teams', 'Lag shot']);
+            for(const zoneId of this.manager_.validDmZones()) {
+                const location = DeathMatchLocation.getById(zoneId);  
+                dialog.addItem(location.id, location.name, location.hasTeams ? "Yes" : "No", 
+                                location.lagShot ? "Yes": "No", 
+                    (player) => {
+                        this.manager_.goToDmZone(player, zoneId);
```
Comment (RussellLVP) [inline javascript/features/death_match/death_match_commands.js]: This is now abusable: you can `/deathmatch` and then quickly press enter (to accept the first zone in the dialog) to enter a deathmatch zone without the activity check on line :43, or the teleportation check on line :52.

I would suggest splitting this up in two methods:

```javascript
async onDmCommand(player, zone) {
    if (DeathMatchLocations.hasLocation(zone))
        return enterDeathMatchZone(zone, player);

    // .. show dialog ..
    dialog.addItem(
        'foo', 'bar', 'baz',
        DeathMatchCommands.prototype.enterDeathMatchZone.bind(this, zone));
}

async enterDeathMatchZone(zone, player) {
    if (player.activity …) {}
    if (teleportStatus…) {}

    this.manager_.goToDmZone(player, zone);
}
```

Why reverse the argument order for `enterDeathMatchZone`? Because it means that you can use the `bind` function without having to create a listener, like you do now 😃

File: javascript/features/death_match/death_match_manager.js [hunk 7]
```diff
@@ -29,38 +29,20 @@ export class DeathMatchManger {
 
         this.callbacks_ = new ScopedCallbacks();
         this.callbacks_.addEventListener(
-            'playerresolveddeath', DeathMatchManger.prototype.onPlayerDeath.bind(this));
-        this.callbacks_.addEventListener(
-            'playerspawn', DeathMatchManger.prototype.onPlayerSpawn.bind(this));
-        this.callbacks_.addEventListener(
-            'playerdisconnect', DeathMatchManger.prototype.onPlayerDisconnect.bind(this));
+            'playerdeath', DeathMatchManger.prototype.onPlayerDeath.bind(this));
```
Comment (RussellLVP) [inline javascript/features/death_match/death_match_manager.js]: I think you really want to keep using `playerresolveddeath` here. Why did you change it back?

The difference is that we have a bunch of abuse and hammering checks that happen before `playerresolveddeath` is called. It's also faster, because we don't block the server on it.

Comment (RussellLVP) [reply javascript/features/death_match/death_match_manager.js]: Actually, I see that the `MockPlayer.prototype.die` function doesn't account for that yet. That's a bug, which I fixed:

https://github.com/LVPlayground/playground/commit/c574e78686f084a2b62a24dde53bc03836a1e168

**Review by RussellLVP — APPROVED** [review]
Approved, please address the new abuse vector before landing :)
